### PR TITLE
refactor: directly integrate redux into react flow

### PR DESCRIPTION
**Pull request by nguyenvukhang**
React flow now literally reads nodes and edges directly from redux. No more manual state-syncing.

Specifically, redux's graph's flowNodes and flowEdges.

### Other changes
- Each redux slice is completely unaware of each other.
  - So now there is one slice at `@/store/modtree` that manages all three of `user`, `degree`, and `graph`.
- Display `prereqTree` in module info modal for quick debug access.
- Retire `api` in frontend in favour of simple functions wrapping `redux` dispatches

### CODE REVIEW

File: apps/web/components/flow/pane.tsx [hunk 1]
```diff
@@ -1,127 +1,61 @@
...
+    // update redux state
+    dispatch(setNodesAndEdges(flowNodes.map(reselect)))
   }
 
   return (
     <ReactFlow
-      nodes={nodes}
-      edges={edges}
+      nodes={flowNodes}
+      edges={flowEdges}
```
Comment (nguyenvukhang) [inline apps/web/components/flow/pane.tsx:47]: directly integrate redux with react-flow.
no more retarded useEffect calls just to sync redux state with react flow state.

File: apps/web/components/graph-title.tsx [hunk 2]
```diff
@@ -1,43 +1,19 @@
...
-  const [graphs, setGraphs] = useState<ModtreeApiResponse.Graph[]>([])
-  // placeholder state
-  // updated correctly in useEffect
-  const state = {
-    graph: useState<ModtreeApiResponse.Graph>(graphs[0]),
-  }
-
-  useEffect(() => {
-    trpc.query('graphs', graphIds).then((graphs) => {
-      setGraphs(graphs)
```
Comment (nguyenvukhang) [inline apps/web/components/graph-title.tsx:22]: God bless this database query

File: apps/web/components/graph-title.tsx [hunk 3]
```diff
@@ -1,43 +1,19 @@
...
-      // mainGraph is different from graphs
-      // mainGraph has an extra property selectedCodes
-      // removing that property is still not enough for React to check equality
-      const mainGraph = store.getState().graph
-      const graph = graphs.find((g) => g.id === mainGraph.id)
-      if (graph) state.graph[1](graph)
-    })
-  }, [graphIds])
```
Comment (nguyenvukhang) [inline apps/web/components/graph-title.tsx:30]: 🤯

File: apps/web/components/modals/module-info/contents.tsx [hunk 4]
```diff
@@ -16,62 +15,15 @@ export function ModuleDetails() {
...
-      .then((g) => {
-        setGraph(g)
-        // 5. Dispatch new nodes
-        dispatch(
-          setFlow({
-            flowNodes: nodes,
-            flowEdges: edges,
-          })
-        )
-      })
```
Comment (nguyenvukhang) [inline apps/web/components/modals/module-info/contents.tsx:73]: 🤩

File: apps/web/components/modals/module-info/contents.tsx [hunk 5]
```diff
@@ -90,6 +42,9 @@ export function ModuleDetails() {
           <Button onClick={handleAddButton}>Add to graph</Button>
         </div>
       )}
+      {devEnv && (
+        <pre>{JSON.stringify(module.prereqTree || 'no pre-reqs', null, 2)}</pre>
+      )}
```
Comment (nguyenvukhang) [inline apps/web/components/modals/module-info/contents.tsx:47]: feat: show pre-req tree in module modal for quick reference while debugging

File: libs/repos/src/graph/repo.ts [hunk 6]
```diff
@@ -201,6 +201,7 @@ export class GraphRepository extends BaseRepo<Graph> {
   ): Promise<Graph> {
     return this.save({
       ...graph,
+      modulesPlaced: props.flowNodes.map((n) => n.data),
```
Comment (nguyenvukhang) [inline libs/repos/src/graph/repo.ts:204]: temporary solution to keep `modulesPlaced` in sync with `flowNodes`

**Review by weiseng18 — APPROVED** [review]

**Pull request merged**